### DISCUSSION TITLE
Does current NNUE correctly evaluate 7 piece endgames?

### DISCUSSION THREAD

**Discussion by RogerThiede** [official-stockfish/Stockfish#4037] [Q&A]
Tablebase WDL are well known for 7 piece endgames.  Does the current NNUE correctly evaluate the best winning (or drawing) move at depth 1 for every given 7 piece endgame?

If it does not, what about 6 piece endgames, 5 piece endgames, and so on?

If it does not, what is the minimum depth where it does correctly evaluate the best winning (or drawing) move for every given 7 piece endgame?

**Comment by ghost** [comment]
> If it does not, what is the minimum depth where it does correctly evaluate the best winning (or drawing) move for every given 7 piece endgame?

the minimum depth for evaluate best winning is 22

**Reply by niklasf** [reply]
Definitely incorrect (disproven by many high-DTZ positions). Given the number of 7 or even just 6 piece positions, I don't think anyone can answer this question today. Analysis for lower piece counts might be feasible.

**Reply by RogerThiede** [reply]
Can you help me understand why _only_ lower piece counts might be feasible to calculate?  I must be missing something fundamental.

Please correct this sentence if I'm wrong: The work required to calculate the 7 piece WDL Tablebase (even if given a known 6 piece WDL Tablebase to assist,) is at least a magnitude more work than to execute the NNUE evaluation for each position in an already computed 7 piece WDL Tablebase and compare if it's correct or not.  Since calculating the 7 piece WDL Tablebase has been proven feasible, the later should also be feasible, if not simply for the fact that what I'm asking about is O(1).

I'm curious about determining what number of pieces that the current NNUE correctly evaluates the best winning (or drawing) move at depth 1, especially if no-one has invested time into this.  I'm expecting it at least plays perfectly with 3 piece positions, where I use the term perfectly to simply mean the highest (or tied) move evaluation reaches a winning 7 piece WDL position if possible, or else it reaches a drawing 7 piece WDL position if the best possible is a forced draw, and I don't care about feeding it a losing position.  Is there a better word for this than "_perfectly_"?  There may be minutia regarding _perfectly_ referring to the fastest move to win which shouldn't matter as long as it still leads to a legal win.

**Reply by niklasf** [reply]
For depth 1:

> The work required to calculate the 7 piece WDL Tablebase (even if given a known 6 piece WDL Tablebase to assist,) is at least a magnitude more work than to execute the NNUE evaluation for each position in an already computed 7 piece WDL Tablebase and compare if it's correct or not.

It depends. First, it might require some expert knowledge to create tools to scan the tablebases efficiently, rather than naively using point lookups. Next, I assume that the efficient iteration order over the tablebases is not necessarily compatible with efficient updates of the NNUE evaluation, so computing evals might be more expensive than usual. But really, I guess the important part is

> compare if it's correct or not

Checking if a "correct" move is found, would require probing child positions, blowing up the complexity. More efficient would be checking if there's a global threshold such that all evaluations under/over are drawn/winning respectively. (But that would be a stronger criterium ... usually it suffices that evaluations locally guide play in the right direction).

The good news would be that (unlike generating tablebases) all of this is massively parallelizable, and does not require large amounts of RAM on singular machines.

For arbitrary depths:

Going back to your original question, I'd expect the biggest obstacle to be depth. A comparatively small number, but time taken to reach higher depths is exponential. Suppose you have not yet found a counter example to movepicker's claim about depth 22. Then you need to search each position to depth 22, including positions where lower depths are (perhaps randomly) correct, because higher depths might regress. I am sure there are 7-piece positions where it takes seconds (if not minutes, hours, or forever, depending on your stance on the 50-move rule) to find a correct move. So you'd have to spend on the order of seconds (minutes, hours, ...) on each 7-piece position, to check all of them at the same depth.

**Reply by ghost** [reply]
> For depth 1

_depth 1 does not make sense when is Quiescence search_
